descriptors: taproot partial descriptors

Eunovo opened the pull request:
This PR adds support for rawnode() and rawleaf() under tr() descriptor discussed in #24114 and [tr-rawnode-and-rawleaf](https://delvingbitcoin.org/t/tr-rawnode-and-rawleaf-support/901)

Eunovo committed:
descriptor: Add TaprootNode

Prepare TRDescriptor to add omitted branches and leaves with user-provided version

Eunovo committed:
descriptor: Implement rawnode() and rawleaf() partial descriptor

ajtowns commented on src/script/descriptor.cpp:
```diff
@@ -1087,17 +1087,36 @@ class WSHDescriptor final : public DescriptorImpl
+            if (m_nodes[pos].type == TRNodeType::NODE_HASH) {
+                builder.AddOmitted(m_nodes[pos].depth, uint256(Span(scripts[pos])));
+            } else if (m_nodes[pos].type == TRNodeType::LEAF_SCRIPT) {
+                builder.Add(m_nodes[pos].depth, scripts[pos], m_nodes[pos].leaf_version.value());
```
Having an `optional` when in one branch you ignore the value and in the other you assume it always has a value doesn't make a lot of sense to me.

Would it make sense to drop `leaf_version` from `TRNodeInfo` and store the leaf version as the first byte of the script? Alternatively, could just drop the optional and default it to 0.

Eunovo replied:
> Having an `optional` when in one branch you ignore the value and in the other you assume it always has a value doesn't make a lot of sense to me.
> 
> Would it make sense to drop `leaf_version` from `TRNodeInfo` and store the leaf version as the first byte of the script? Alternatively, could just drop the optional and default it to 0.

@ajtowns I prefer to drop the optional and default leaf_version to `0`

ajtowns reviewed (commented):
Seems okay to me. Presumably should update doc/descriptors.md though, and might also be nice to have some functional tests demonstrating it works end-to-end? (Add some examples to `DESCS` in wallet_miniscript.py maybe?

Eunovo commented:
> Seems okay to me. Presumably should update doc/descriptors.md though, and might also be nice to have some functional tests demonstrating it works end-to-end? (Add some examples to `DESCS` in wallet_miniscript.py maybe?

@ajtowns Will do!

Eunovo committed:
test: add e2e tests for rawnode() and rawleaf()

Eunovo committed:
doc: Add reference for rawnode() and rawleaf()

Eunovo commented:
@ajtowns Thanks for the review. I added some examples to `wallet_miniscript.py` and updated `descriptors.md`

dergoegge commented:
Found a crash in the `mocked_descriptor_parse` fuzz target, to reproduce:

```
$ echo "dHIoJUJELHJhd2xlYWYoQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCLEE3LCUyNywlQjcsJTIzLCVCZCwlRkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQmNGKSk=" | base64 --decode > pr30243-mocked_descriptor_parse.crash
$ FUZZ=mocked_descriptor_parse src/test/fuzz/fuzz pr30243-mocked_descriptor_parse.crash
fuzz: script/signingprovider.cpp:368: TaprootBuilder &TaprootBuilder::Add(int, Span<const unsigned char>, int, bool): Assertion `(leaf_version & ~TAPROOT_LEAF_MASK) == 0' failed.
```

Eunovo commented:
> Found a crash in the `mocked_descriptor_parse` fuzz target, to reproduce:
> 

Thanks for this @dergoegge I'll fix it

Eunovo commented:
> Found a crash in the `mocked_descriptor_parse` fuzz target, to reproduce:
> 
> ```
> $ echo "dHIoJUJELHJhd2xlYWYoQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCLEE3LCUyNywlQjcsJTIzLCVCZCwlRkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQmNGKSk=" | base64 --decode > pr30243-mocked_descriptor_parse.crash
> $ FUZZ=mocked_descriptor_parse src/test/fuzz/fuzz pr30243-mocked_descriptor_parse.crash
> fuzz: script/signingprovider.cpp:368: TaprootBuilder &TaprootBuilder::Add(int, Span<const unsigned char>, int, bool): Assertion `(leaf_version & ~TAPROOT_LEAF_MASK) == 0' failed.
> ```

@dergoegge Pushed https://github.com/bitcoin/bitcoin/pull/30243/commits/7efb115fd10584157d3ad752fee7e8ee4d267b02 to fix this

brunoerg commented on src/script/descriptor.cpp:
```diff
@@ -1886,6 +1915,42 @@ std::unique_ptr<DescriptorImpl> ParseScript(uint32_t& key_exp_index, Span<const
+            return nullptr;
+        }
+        if (leaf_version->size() == 0) {
+            error = "Expected Leaf Version but not provided";
```
In 6908349fa559c16164c76fff2f140a39f52279e2 "descriptor: Implement rawleaf(() partial descriptor": Please correct me if I'm wrong, but is this check needed? If the expected leaf version is not provided, won't it be verified before (in `if (!Const(",", expr))`?.

Eunovo replied:
Thanks for the review @brunoerg. I pushed https://github.com/bitcoin/bitcoin/pull/30243/commits/abe28609c0c7124c8ecf3f09adb537e3b642d50c which adds a test that triggers this error

brunoerg replied:
Cool

brunoerg commented on src/test/descriptor_tests.cpp:
```diff
@@ -643,6 +643,25 @@ BOOST_AUTO_TEST_CASE(descriptor_test)
     Check("tr(a34b99f22c790c4e36b2b3c2c35a36db06226e41c692fc82b8b56ac1c540c5bd,{and_v(and_v(v:hash256(ae253ca2a54debcac7ecf414f6734f48c56421a08bb59182ff9f39a6fffdb588),v:pk(KykUPmR5967F4URzMUeCv9kNMU9CNRWycrPmx3ZvfkWoQLabbimL)),older(42)),multi_a(2,adf586a32ad4b0674a86022b000348b681b4c97a811f67eefe4a6e066e55080c,KztMyyi1pXUtuZfJSB7JzVdmJMAz7wfGVFoSRUR5CVZxXxULXuGR)})", "tr(a34b99f22c790c4e36b2b3c ...
     Check("tr(a34b99f22c790c4e36b2b3c2c35a36db06226e41c692fc82b8b56ac1c540c5bd,{and_v(and_v(v:hash256(ae253ca2a54debcac7ecf414f6734f48c56421a08bb59182ff9f39a6fffdb588),v:pk(KykUPmR5967F4URzMUeCv9kNMU9CNRWycrPmx3ZvfkWoQLabbimL)),older(42)),multi_a(2,adf586a32ad4b0674a86022b000348b681b4c97a811f67eefe4a6e066e55080c,KztMyyi1pXUtuZfJSB7JzVdmJMAz7wfGVFoSRUR5CVZxXxULXuGR)})", "tr(a34b99f22c790c4e36b2b3c ...
 
+    // Can have rawwnode under tr()
```
```suggestion
    // Can have rawnode under tr()
```

brunoerg commented:
You can squash some commits.

Eunovo commented:
> You can squash some commits.

Thanks @brunoerg. Done!

brunoerg commented on src/script/descriptor.cpp:
```diff
@@ -1822,6 +1900,63 @@ std::unique_ptr<DescriptorImpl> ParseScript(uint32_t& key_exp_index, Span<const
     }
+    if (ctx == ParseScriptContext::P2TR && Func("rawnode", expr)) {
+        std::string str(expr.begin(), expr.end());
+        if (!IsHex(str)) {
```
In 21b6bf4268e40020523d027e2811402b17b755eb "descriptor: Implement rawnode() partial descriptor": Is this `IsHex` verification necessary? Wouldn't it be verified in `ParseHex`? 

Anyway, a test would be nice.

Eunovo replied:
In https://github.com/bitcoin/bitcoin/pull/30243/commits/0b30b690e7067240b05ece17fdc7ad44549368a3 I now use `TryParseHex` and just check for the presence of a response

brunoerg commented on doc/descriptors.md:
```diff
@@ -83,6 +83,8 @@ Descriptors consist of several types of expressions. The top level expression is
 - `raw(HEX)` (top level only): the script whose hex encoding is HEX.
 - `rawtr(KEY)` (top level only): P2TR output with the specified key as output key. NOTE: while it's possible to use this to construct wallets, it has several downsides, like being unable to prove no hidden script path exists. Use at your own risk.
+- `rawnode(HEX)` (inside `tr` only): specify tree or branch merkle root in HEX encoding. NOTE: this hides script paths which are only revealed after spending. Use at your own risk
+- `rawleaf(HEX,[LEAF_VERSION_HEX])`(inside `tr` only): specify raw leaf script and leaf version in hex encoding. `LEAF_VERSION_HEX` defaults to `c0`
```
In a945bc21d4ad4fd5da68ae783ec295d27f50227a "doc: Add reference for rawnode() and rawleaf()": Wouldn't it be `rawleaf(HEX[,LEAF_VERSION_HEX])`?

brunoerg replied:
maybe
```suggestion
- `rawleaf(HEX,[LEAF_VERSION_HEX])`(inside `tr` only): specify raw leaf script and leaf version in HEX encoding. `LEAF_VERSION_HEX` defaults to `c0`
```

Eunovo commented on src/script/descriptor.cpp:
```diff
@@ -1087,17 +1087,36 @@ class WSHDescriptor final : public DescriptorImpl
+/** A struct hold information on a node taproot descriptor script tree */
+struct TRNodeInfo {
+    int depth;
+    int leaf_version;
```
No it doesn't. I have changed the type to `uint8_t`

Eunovo commented:
- Rebased on top of master https://github.com/bitcoin/bitcoin/commit/a865494deeff7dedcad7140299aee00ab3cdd62c
- Added a test for rawnode with multipath descriptors in https://github.com/bitcoin/bitcoin/pull/30243/commits/133b30019c139a3a0e624df2015f65c64040c0c8

achow101 commented:
Needs a bip

ismaelsadeeq commented on src/script/descriptor.cpp:
```diff
@@ -1163,17 +1163,36 @@ class WSHDescriptor final : public DescriptorImpl
+};
+
+/** A struct hold information on a node taproot descriptor script tree */
+struct TRNodeInfo {
```
In  "descriptor: Add TRNodeInfo" 475981527f052f7c9709d1f3a90270770f4dc160

```suggestion
struct TaprootNode {
```

ismaelsadeeq commented on src/script/descriptor.cpp:
```diff
@@ -2039,11 +2059,18 @@ std::vector<std::unique_ptr<DescriptorImpl>> ParseScript(uint32_t& key_exp_index
+                node.depth = depths[pos];
+                node.leaf_version = TAPROOT_LEAF_TAPSCRIPT;
+                node.type = TRNodeType::LEAF_SCRIPT;
+                this_nodes.emplace_back(node);
```
In  "descriptor: Add TRNodeInfo" 475981527f052f7c9709d1f3a90270770f4dc160

You should add a constructor to `TRNodeInfo`, this will prevent constructing the object and individually adding all elements before emplacing to the vector here and other places.

```suggestion
                this_nodes.emplace_back(depths[pos], TAPROOT_LEAF_TAPSCRIPT, TRNodeType::LEAF_SCRIPT);
```

ismaelsadeeq commented on src/script/descriptor.cpp:
```diff
@@ -1418,6 +1420,59 @@ class RawTRDescriptor final : public DescriptorImpl
+public:
+    RawNodeDescriptor(std::vector<unsigned char> bytes) : DescriptorImpl({}, "rawnode"), m_bytes(std::move(bytes)) {}
+
+    bool IsSolvable() const final { return false; }
+
+    bool IsSingleType() const final { return true; }
```
In "descriptor: Implement rawnode() partial descriptor" a9a9e78c3f839eceebb3abd473d8658fabd92f1f
This can be inlined?

ismaelsadeeq commented on src/script/descriptor.cpp:
```diff
@@ -1418,6 +1420,59 @@ class RawTRDescriptor final : public DescriptorImpl
+public:
+    RawLeafDescriptor(CScript leaf_script, uint8_t leaf_version) : DescriptorImpl({}, "rawleaf"), m_leaf_script(leaf_script), m_leaf_version(leaf_version) {}
+
+    bool IsSolvable() const final { return false; }
```
In  "descriptor: Implement rawnode() partial descriptor" a9a9e78c3f839eceebb3abd473d8658fabd92f1f
We are duplicating this code here!

I wonder if `DescriptorImpl` will just define this duplicate methods defined with the truthy or falsy values,  All descriptor that require concrete implementation will just override those methods.

ismaelsadeeq reviewed (commented):
> Needs a bip

There is a BIP draft opened @furszy that define these output descriptors https://github.com/bitcoin/bips/pull/1721

But the BIP and PR diverge in `rawleaf` output descriptor, where the bip is omitting the leaf version, I prefer the approach taken in this PR which is also in line with the initial issue discussion in the issue.

Eunovo commented:
Rebased on master https://github.com/bitcoin/bitcoin/commit/770d39a37652d40885533fecce37e9f71cc0d051
- changed `Span` to `std::span`

Eunovo commented:
Rebased on master.
